### PR TITLE
#71 set up authentication

### PR DESCRIPTION
**Pull request by oneroyalace**
#71 
This PR sets up basic user authentication for Zenodotus. It uses Devise to create a User model, allowing visitors to create accounts and log in to the app. In fact, as currently setup, it forces user to log in -- attempts to access any controller without logging in will fail + reroute to the login page. 

It uses a boolean flag to mark User accounts as admins. For now, admins have the special privileges of viewing their API key ( which can be used to authenticate remote requests) and approving/denying signup requests.

### CODE REVIEW

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -3,12 +3,30 @@ class ApplicationController < ActionController::Base
...
+
+  sig { void }
+  def authenticate_user_from_api_key!
+    api_key = params[:api_key].presence
+    hashed_api_key = Digest::SHA256.hexdigest(api_key)
+    api_key = hashed_api_key && ApiKey.find_by_hashed_api_key(hashed_api_key)
+
+    # Devise.secure_compare prevents timing attacks
+    if api_key && Devise.secure_compare(api_key.hashed_api_key, hashed_api_key)
+      sign_in(api_key.user, store: false)
```
Comment (oneroyalace) [inline app/controllers/application_controller.rb:23]: @cguess Would this create a session, allowing the user to send later request without the API Key?

Comment (oneroyalace) [reply app/controllers/application_controller.rb:23]: Ah, never mind. The authentication is run each time before the user's request touches a controller

File: config/environments/development.rb [hunk 2]
```diff
@@ -49,6 +49,11 @@
 
   config.action_mailer.perform_caching = false
 
+  config.action_mailer.default_url_options = { host: 'localhost', port: 3000 }
+
+  config.action_mailer.delivery_method = :letter_opener
```
Comment (oneroyalace) [inline config/environments/development.rb:54]: We should probably take this out now

Comment (cguess) [reply config/environments/development.rb]: Good call, fixed.

File: app/controllers/application_controller.rb [hunk 3]
```diff
@@ -3,21 +3,38 @@ class ApplicationController < ActionController::Base
   extend T::Sig
   extend T::Helpers
 
-  # Return a class that can handle a given +url+
-  sig { params(url: String).returns(T.nilable(Class)) }
-  def model_for_url(url)
```
[the remark this answers is not in the document]
Comment (oneroyalace) [reply app/controllers/application_controller.rb:8]: Logs from an archive action: 
```
Started POST "/archive/add" for ::1 at 2021-08-16 10:36:47 -0400
Processing by ArchiveController#submit_url as TURBO_STREAM
  Parameters: {"authenticity_token"=>"[FILTERED]", "url_to_archive"=>"https://twitter.com/kairyssdal/status/1415029747826905090", "commit"=>"Archive"}
Completed 500 Internal Server Error in 8ms (ActiveRecord: 0.0ms | Allocations: 4166)

NoMethodError (undefined method `model_for_url' for #<ArchiveController:0x0000000000dc78>)
```

File: app/controllers/archive_controller.rb [hunk 4]
```diff
@@ -21,6 +33,10 @@ class SubmitUrlParams < T::Struct
   # @params {url_to_archive} the url to pull in
   sig { void }
   def submit_url
+    # unless user_signed_in?
+    #   format.html { redirect_to :new_user_session_path }
+    #   return
+    # end
     typed_params = TypedParams[SubmitUrlParams].new.extract!(params)
     url = typed_params.url_to_archive
     object_model = model_for_url(url)
```
Comment (cguess) [inline app/controllers/archive_controller.rb:42]: This will break, write a test

**Comment by oneroyalace** [comment]
All tests now passing locally

**Review by cguess — APPROVED** [review]

**Pull request merged**